crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+            );
+        }
+
+        let checkpoint = circuit.commit().unwrap();
+        circuit.kill().unwrap();
+        super::INTERNED_STRING_CACHE.clear();
+
+        let (mut circuit, (_hinput_strings, hqueries, houtput_strings)) =
+            interner_test_circuit(path.as_path(), Some(checkpoint.uuid));
+
+        query(
+            &mut circuit,
+            &hqueries,
+            &houtput_strings,
+            (0..1_000)
+                .map(|i| i.to_string())
+                .collect::<Vec<_>>()
+                .iter()
+                .map(|s| s.as_str()),
+        );
+
+        assert_eq!(super::INTERNED_STRING_CACHE.len(), 1000);
+        assert!(super::INTERNED_STRING_CACHE.misses() - old_missed <= 10000);
+        circuit.kill().unwrap();
+    }
+
+    /// Insert and then delete some strings.
+    /// INTERNED_STRING_BY_ID should be empty in the end.
+    #[test]
+    #[serial]
+    fn test_interner_deletions() {
+        let path = tempfile::tempdir().unwrap().keep();
+
+        let (mut circuit, (hinput_strings, hqueries, houtput_strings)) =
+            interner_test_circuit(path.as_path(), None);
+
+        for batch in 0..1_000 {
+            let values = (batch * 100..(batch + 1) * 100)
+                .map(|i| i.to_string())
+                .collect::<Vec<_>>();
+            let mut chunk = values
+                .iter()
+                .map(|i| Tup2(SqlString::from(i.as_str()), 1))
+                .collect::<Vec<_>>();
+            hinput_strings.append(&mut chunk);
+
+            query(
+                &mut circuit,
+                &hqueries,
+                &houtput_strings,
+                values.iter().map(String::as_str),
+            );
+        }
+
+        let checkpoint = circuit.commit().unwrap();
+        circuit.kill().unwrap();
+        super::INTERNED_STRING_CACHE.clear();
+
+        let (mut circuit, (hinput_strings, _hqueries, _houtput_strings)) =
+            interner_test_circuit(path.as_path(), Some(checkpoint.uuid));
+
+        for batch in 0..1_000 {
+            let values = (batch * 100..(batch + 1) * 100)
+                .map(|i| i.to_string())
+                .collect::<Vec<_>>();
+            let mut chunk = values
+                .iter()
+                .map(|i| Tup2(SqlString::from(i.as_str()), -1))
+                .collect::<Vec<_>>();
+            hinput_strings.append(&mut chunk);
+            circuit.step().unwrap();
+        }
+        circuit.step().unwrap();
+        circuit.step().unwrap();
+
+        let mut cursor = super::INTERNED_STRING_BY_ID.read().unwrap().cursor();
+        while cursor.key_valid() {
+            while cursor.val_valid() {
+                assert_eq!(**cursor.weight(), 0);
+                //println!("weight: {}", **cursor.weight());
+                cursor.step_val();
+            }
+            cursor.step_key();
+        }
+        circuit.kill().unwrap();
+    }
+
+    #[test]
+    #[serial]
+    fn test_interner_bulk() {
```

Comment:
does any of these tests get to exercise evictions from the cache?

Reply:
They don't (but I did test it from SQL). I'll add such a test after making cache size configurable.